A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,236 @@
+
+* [gRFC A42: xDS Ring Hash LB Policy][A42]
+* [gRFC A61: IPv4 and IPv6 Dualstack Backend Support][A61]
+* [gRFC A74: xDS Config Tears][A74]
+
+## Proposal
+
+### Explicitly setting the request hash key
+
+A new field `request_hash_header` will be added to the `ring_hash` policy
+config:
+
+```proto
+    message RingHashLoadBalancingConfig {
+      // (existing fields omitted)
+      string request_hash_header = 3;
+    }
+```
+
+Upon loading the load balancing config, if the `request_hash_header` field
+contains a value that is not a valid header name, then the configuration is
+rejected. If the `request_hash_header` refers to a binary header (suffixed with
+`-bin`), the configuration is also rejected.
+
+At pick time:
+- If `request_hash_header` is empty, then the request hash key will be based on
+the xDS hash policy in RDS, which keeps the existing LB configuration for ring
+hash working as before with xDS.
+- If `request_hash_header` is not empty, and the header has a non-empty value,
+then the request hash key will be set to this value. If the header contains
+multiple values, then values are joined with a comma `,` character before
+hashing.
+- If `request_hash_header` is not empty, and the request has no value associated
+with the header or its value is empty, then the picker will generate a random
+hash for the request. If the use of this random hash triggers a connection
+attempt (according to the rules described in [A42: Picker
+Behavior][A42-picker-behavior] and updated in [A61: Ring Hash][A61-ring-hash]),
+then before queuing the pick, the picker will scan forward searching for a
+subchannel in `READY` state. If it finds a subchannel in `READY` state, the
+picker returns it.
+
+The following pseudo code describes the updated picker logic:
+
+```
+// Determine request hash.
+using_random_hash = false;
+if (config.request_hash_header.empty()) {
+  request_hash = call_attributes.hash;
+} else {
+  header = headers.find(config.request_hash_header);
+  if (header == null) {
+    using_random_hash = true;
+    request_hash = ComputeRandomHash();
+  } else {
+    request_hash = ComputeHash(header);
+  }
+}
+
+// Do pick based on hash.
+first_index = ring.FindIndexForHash(request_hash);
+requested_connection = false;
+for (i = 0; i < ring.size(); ++i) {
+  index = (first_index + i) % ring.size();
+  if (ring[index].state == READY) {
+    return ring[index].picker->Pick(...);
+  }
+  if (requested_connection) continue;
+  if (ring[index].state == IDLE) {
+    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
+    if (using_random_hash) {
+      requested_connection = true;
+      continue;
+    }
+    return PICK_QUEUE;
+  }
+  if (ring[index].state == CONNECTING) {
+    return PICK_QUEUE;
+  }
+}
+if (requested_connection) return PICK_QUEUE;
+return PICK_FAIL;
+```
+
+This behavior ensures that a single request creates at most one new connection,
+and that a request missing the header does not add extra latency in the common
```

Comment:
Thanks Eric for pointing out this flaw, and Mark for providing the updated pseudo code.

> But I also wonder if we want to take a completely alternate approach.

Alternative approaches were discussed in https://github.com/grpc/proposal/pull/412#discussion_r1462445670.
1. Use something like config selector to set the hash outside the picker, but we don't have a cross language API for this yet.
2. Make it possible for the picker to retrieve the hash across picks for a given request. Again IIUC we don't have an API for this.
3. What we chose here -- make sure we don't cause more than one connection attempt per request by taking the ring state into account.
4. Fail the request
5. Use a hardcoded hash.

There are details on the comment thread about the pros and cons of each approach. I can't really speak for 1 or 2, they are more work which may be worth it. I could live with 4 and 5, although 3 is arguably nicer in terms of behavior, as no hash values results in a random load distribution.

> So I think what we want is something like this: ...

In the code you propose, we end up failing picks when all children are in either `CONNECTING` or `TRANSIENT_FAILURE` state, even if this request actually caused one of the child to transition to `CONNECTING`. That doesn't seem right to me. Shouldn't we queue the pick in that case? I would have expected something like:

```
  // If we didn't find CONNECTING but found IDLE, start a connection attempt.
  if (!found_connecting && found_idle_index != -1) {
    ring[found_idle_index].endpoint.TriggerConnectionAttemptInControlPlane();
    if (found_ready_index == -1) return PICK_QUEUE;
  }
  // If we found READY, use it.
  if (found_ready_index != -1) {
    return ring[found_ready_index].picker->Pick(...);
  }  
  // If we found CONNECTING, queue the pick.
  if (found_connecting) {
    return PICK_QUEUE;
  }
  // Didn't find READY or CONNECTING, and didn't start a connection attempt.
  return PICK_FAIL;
```

But then, this would mean that multiple requests using a random hash being issued concurrently can only trigger one connection at a time. This behavior may be acceptable, but that's not my understanding of what Eric was suggesting. IIUC Eric suggested using the copy of the previous children states that we already have in ring hash ([here in c-core](https://github.com/grpc/grpc/blob/b1896a2136e9f60e026daeaaf393cb8cb143be72/src/core/load_balancing/ring_hash/ring_hash.cc#L212)), and compare it to the new state to avoid updating the picker in the case where the state update is due to a subchannel going from `IDLE` to `CONNECTING`. This would effectively avoid triggering picks as a result of triggering a connection attempt.

Reply:
I need to catch up on the conversation, but we discussed it on Friday (after Mark's message here) and seems it didn't make it here yet. I'll be in meetings all day to day, so want to get it out now.

We came up with initializing `requested_connection` specially in the random hash case (note that it also has the change to CONNECTING that was mentioned in my other comment). When the picker is created, it'd go through the subchannels and compute `picker_has_a_subchannel_connecting`, and then use it at picking time:

```c++
first_index = ring.FindIndexForHash(request_hash);
requested_connection = using_random_hash && picker_has_a_subchannel_connecting;
for (i = 0; i < ring.size(); ++i) {
  index = (first_index + i) % ring.size();
  if (ring[index].state == READY) {
    return ring[index].picker->Pick(...);
  }
  if (requested_connection) continue;
  if (ring[index].state == IDLE) {
    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
    if (using_random_hash) {
      requested_connection = true;
      continue;
    }
    return PICK_QUEUE;
  }
  if (ring[index].state == CONNECTING) {
    if (using_random_hash) {
      continue;
    }
    return PICK_QUEUE;
  }
}
if (requested_connection) return PICK_QUEUE;
return PICK_FAIL;
```

Reply:
> We already share child state outside of the picker, so...

This comment of mine was wrong. So avoiding updating the Picker is probably a mess.

> 2. Make it possible for the picker to retrieve the hash across picks for a given request. Again IIUC we don't have an API for this.

We talked about this option a lot. We _love_ how simple [ring_hash's picker is since A61](https://github.com/grpc/proposal/blob/master/A61-IPv4-IPv6-dualstack-backends.md#ring-hash). And our initial pseudocode to fix this just within ring_hash was really, really ugly.

But I in particular was worried about adding this API because there's multiple architectural choices to be made, and each choice could impact whether it is useful for other cases. Needing to store state has come up multiple times, but without looking at multiple cases and confirming the design works for them, it seemed _unlikely_ that it would happen to be helpful for other cases.

Notably, in general we'd probably need to use the LB storing state as part of the key for the state. Yes, ring_hash is okay if the hash is shared across leafs, but that isn't generally true. (The concern is if a later pick takes a different path through the LB tree; we don't want the LB to be confused by the state.)

> We came up with initializing requested_connection specially in the random hash case

This caused the pseudocode to greatly simplify, such that it did seem an acceptable complexity in ring_hash to avoid moving the complexity into our architecture. That pseudocode I shared was a first cut; we were going to look at other simplifications, but it was clean enough to be acceptable.

One thing that might be preferable, is to just use a separate loop/function for the `using_random_hash` case to avoid all the ifs. This is the equivalent to what I posted Monday, but only for the `use_random_hash=true` case:
```c++
first_index = ring.FindIndexForHash(request_hash);
requested_connection = picker_has_a_subchannel_connecting;
for (i = 0; i < ring.size(); ++i) {
  index = (first_index + i) % ring.size();
  if (ring[index].state == READY) {
    return ring[index].picker->Pick(...);
  }
  if (requested_connection) continue;
  if (ring[index].state == IDLE) {
    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
    requested_connection = true;
    continue;
  }
}
if (requested_connection) return PICK_QUEUE;
return PICK_FAIL;
```

I think we'd consider them essentially equivalent, so feel free to put the combined or the separate psuedocode in the gRFC; whatever your preference. That's assuming you like this pseudocode at all.